hathor/execution_manager.py

```diff
@@ -0,0 +1,51 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import os
+from typing import NoReturn
+
+from structlog import get_logger
+
+from hathor.event import EventManager
+from hathor.transaction.storage import TransactionStorage
+
+logger = get_logger()
+
+
+class ExecutionManager:
+    """Class to manage actions related to full node execution."""
+    __slots__ = ('_log', '_tx_storage', '_event_manager')
+
+    def __init__(self, *, tx_storage: TransactionStorage, event_manager: EventManager) -> None:
+        self._log = logger.new()
+        self._tx_storage = tx_storage
+        self._event_manager = event_manager
+
+    def crash_and_exit(self, *, reason: str) -> NoReturn:
+        """
+        Calling this function is a very extreme thing to do, so be careful. It should only be called when a
+        critical, unrecoverable failure happens. It crashes and exits the full node, rendering the database
+        corrupted, and requiring manual intervention. In other words, a restart with a clean database (from scratch
+        or a snapshot) will be required.
+        """
+        self._tx_storage.full_node_crashed()
+        self._event_manager.full_node_crashed()
+        self._log.critical(
+            'Critical failure occurred, causing the full node to halt execution. Manual intervention is required.',
+            reason=reason,
+            exc_info=True
+        )
+        # We use os._exit() instead of sys.exit() or any other approaches because this is the only one Twisted
+        # doesn't catch.
+        os._exit(-1)
```

Comment:
I prefer to crash using `reactor.crash()` or `sys.exit()` which would allow flushes and other activities to happen.

Reply:
Initially my idea was to purposefully halt execution _immediately_, preventing any other actions such as pending `Deferred`s getting executed. They pollute the logs with unrelated errors, but I guess it's okay. I've changed it in https://github.com/HathorNetwork/hathor-core/pull/809/commits/6978d71f9a18363e78d1d7d4724b512efd84dc73

Reply:
I've seen things like:

```python
try:
    sys.exit(-1)
except:
    os._exit(-1)
```

What do you think?

Reply:
I'm not sure this works, since `sys.exit(-1)` simply raises a `SystemExit(-1)` exception. This would prevent the exception from bubbling up, and would call `os._exit(-1)` every time.

Reply:
Unless that's the intended behavior, which I guess could make sense. The `sys.exit()` would perform its finalization operations, and then we would force the exit with the `os._exit()`, always. Is that your idea?

Reply:
I saw the proposed changelog in another project, but I agree with your remarks on how it would actually work. I guess the current implementation is fine and it should be simple to change this behavior in the future if needed. So we need to move forward the way it is now. Thanks!